controller/lib/hal/hal_stm32.cpp

```diff
@@ -0,0 +1,1008 @@
+// These pins are connected to UART3
+static void InitUARTs()
+{
+// NOTE - The UART functionality hasn't been tested due to lack of hardware!
+//        Need to do that as soon as the boards are available.
+   EnableClock( UART3_BASE );
+
+   GPIO_PinAltFunc( GPIO_B_BASE, 10, 7 );
+   GPIO_PinAltFunc( GPIO_B_BASE, 11, 7 );
+   GPIO_PinAltFunc( GPIO_B_BASE, 13, 7 );
+   GPIO_PinAltFunc( GPIO_B_BASE, 14, 7 );
+
+   rpUART.Init( UART3_BASE, 115200 );
+
+   EnableInterrupt( INT_VEC_UART3, 3 );
+}
+
+static void UART3_ISR(){
+   rpUART.ISR();
+}
+
+uint16_t HalApi::serialRead(char *buf, uint16_t len) {
+   return rpUART.read( buf, len );
+}
+
+uint16_t HalApi::serialBytesAvailableForRead() {
+   return rpUART.RxFull();
+}
+
+uint16_t HalApi::serialWrite(const char *buf, uint16_t len) {
+   return rpUART.write( buf, len);
+}
+
+uint16_t HalApi::serialBytesAvailableForWrite() {
+   return rpUART.TxFree();
+}
+
+/******************************************************************
+ * Watchdog timer (see chapter 32 of reference manual).
+ *
+ * The watchdog timer will reset the system if it hasn't been
+ * re-initialized within a specific amount of time.  It's used
+ * to catch bugs that would otherwise hang the system.  When
+ * the watchdog is enabled such a bug will reset the system
+ * rather then let it hang indefinitely.
+ *****************************************************************/
+void HalApi::watchdog_init() {
+   Watchdog_Regs *wdog = reinterpret_cast<Watchdog_Regs *>(WATCHDOG_BASE);
+
+   // Enable the watchdog timer by writing the appropriate value to it's key register
+   wdog->key    = 0xCCCC;
+
+   // Enable register access
+   wdog->key    = 0x5555;
+
+   // Set the pre-scaler to 0.  That setting will cause the watchdog
+   // clock to be updated at approximately 8KHz.
+   wdog->prescale = 0;
+
+   // The reload value gives the number of clock cycles before the
+   // watchdog timer times out.  I'll set it to 1000 which gives
+   // us about 250ms before a reset.
+   wdog->reload = 2000;
+
+   // Since the watchdog timer runs off it's own clock which is pretty
+   // slow, it takes a little time for the registers to actually get
+   // updated.  I wait for the status register to go to zero which
+   // means it's done.
+   while( wdog->status ){}
+
+   // Reset the timer.  This also locks the registers again.
+   wdog->key    = 0xAAAA;
+}
+
+// Pet the watchdog so he doesn't bite us.
+void HalApi::watchdog_handler() {
+   Watchdog_Regs *wdog = reinterpret_cast<Watchdog_Regs *>(WATCHDOG_BASE);
+   wdog->key    = 0xAAAA;
+}
+
+// Enable clocks to a specific peripherial.
+// On the STM32 the clocks going to various peripherials on the chip
+// are individually selectable and for the most part disabled on startup.
+// Clocks to the specific peripherials need to be enabled through the
+// RCC (Reset and Clock Controller) module before the peripherial can be
+// used.
+// Pass in the base address of the peripherial to enable it's clock
```

Comment:
its

Reply:
Ugh, I always do that!  Thanks and continue to point out mistakes like that.